toolchain/check/check_unit.cpp

```diff
@@ -511,9 +514,66 @@ auto CheckUnit::CheckRequiredDefinitions() -> void {
   }
 }
 
+auto CheckUnit::CheckPoisonedConcreteImplLookupQueries() -> void {
+  // Impl lookup can generate instructions (via deduce) which we don't use, as
+  // we're only generating diagnostics here, so we catch and discard them.
+  context_.inst_block_stack().Push();
+  auto poisoned_queries =
+      std::exchange(context_.poisoned_concrete_impl_lookup_queries(), {});
+  for (const auto& poison : poisoned_queries) {
+    auto witness_result =
+        EvalLookupSingleImplWitness(context_, poison.loc_id, poison.query,
+                                    poison.non_canonical_query_self_inst_id,
+                                    /*poison_concrete_results=*/false);
+    CARBON_CHECK(witness_result.has_concrete_value());
+    auto found_witness_id = witness_result.concrete_witness();
+    if (found_witness_id != poison.impl_witness) {
+      auto witness_to_impl_id = [&](SemIR::InstId witness_id) {
+        auto table_id = context_.insts()
+                            .GetAs<SemIR::ImplWitness>(witness_id)
+                            .witness_table_id;
+        return context_.insts()
+            .GetAs<SemIR::ImplWitnessTable>(table_id)
+            .impl_id;
+      };
+
+      // We can get the `Impl` from the resulting witness here, which is the
+      // `Impl` that conflicts with the previous poison query.
+      auto bad_impl_id = witness_to_impl_id(found_witness_id);
+      const auto& bad_impl = context_.impls().Get(bad_impl_id);
+
+      auto prev_impl_id = witness_to_impl_id(poison.impl_witness);
+      const auto& prev_impl = context_.impls().Get(prev_impl_id);
+
+      // Both impls will give the same interface name, so doesn't matter which
+      // we used here.
+      auto interface_name_id =
+          context_.interfaces().Get(bad_impl.interface.interface_id).name_id;
+
+      CARBON_DIAGNOSTIC(PoisonedImplLookupConcreteResult, Error,
+                        "found `impl` that would change the result of a "
+                        "previous use of `{0} as {1}`",
+                        InstIdAsRawType, SemIR::NameId);
+      auto builder = emitter_.Build(
+          bad_impl.first_decl_id(), PoisonedImplLookupConcreteResult,
+          poison.query.query_self_inst_id, interface_name_id);
+      CARBON_DIAGNOSTIC(PoisonedImplLookupConcreteResultNoteUse, Note,
+                        "use of impl here");
+      builder.Note(poison.loc_id, PoisonedImplLookupConcreteResultNoteUse);
+      CARBON_DIAGNOSTIC(PoisonedImplLookupConcreteResultNotePreviousImpl, Note,
+                        "previously got result from `impl` here");
```

Comment:
Perhaps:
```suggestion
                        "previous use selected this `impl`");
```

Reply:
Done, with slight change (went with "from `impl` here" still over "this `impl`", which speaks of the location consistently)